### PR TITLE
wip:Support multi control plane

### PR DESCRIPTION
**Pull request by lingsamuel**
/kind feature

Changes:
- [x] fix tests - hope finished
- [x] fix missing InitialSetup in `node add`
- [x] fix/implement `node add --control-plane`
- [x] add flag `--control-planes` to config control planes number: `minikube start --control-planes 2`. `--nodes` will automatically enlarge if `nodes < control-planes` and outputs a warning.
- [x] tag primary control plane: add field `PrimaryControlPlane`
- [x] add IP and primary tag in `status`
- [x] reset control plane node before stop it, make sure primary control plane can start without others control planes
- [x] add some step output
- [x] change KindNet.String() from `CNI` to `KindNet`
- [x] `status` command: stacked control plane node `kubeconfig` should be irrelevant. `kubeconfig` still exist in stacked control plane node but configured at `control-plane.minikube.internal:8443` to generate certificate keys.
- [x] backward compatibility
  - [x] `start` from a stopped old cluster config
  - [x] `node add`, `node stop`, `node start`, `stop` a running old cluster
  - [x] `status` warn user outdated config, provides flag `--update-primary-control-plane-tag` to update profile.
- [x] multiple control planes integration tests

Fixes: #7461, part of #7538

Currently finished:

```bash
minikube start
minikube node add --control-plane # Support multi control plane
minikube stop # make sure stop && start works
minikube start
minikube node stop m02 # make sure node stop && node start works
minikube node start m02
```

Some output changes:
```bash
$ minikube status
minikube
type: Control Plane (Primary)
host: Running
kubelet: Running
apiserver: Running
kubeconfig: Configured
timeToStop: Nonexistent
IP: 192.168.39.13

minikube-m02
type: Control Plane
host: Running
kubelet: Running
apiserver: Running
kubeconfig: Irrelevant
timeToStop: Nonexistent
IP: 192.168.39.252
```

Outdated profile and `status` command output:

```js
// ...
"Nodes": [
        {
            "Name": "",
            "IP": "192.168.39.164",
            "Port": 8443,
            "KubernetesVersion": "v1.19.3",
            "PrimaryControlPlane": false, // <- outdated config is false here
            "ControlPlane": true,
            "Worker": true
        }
    ],
// ...
```

```bash
$ minikube status
There is no primary control plane, set --update-primary-control-plane-tag=true to update profile.
minikube
type: Control Plane
host: Running
kubelet: Running
apiserver: Running
kubeconfig: Irrelevant
timeToStop: Nonexistent
IP: 192.168.39.164
```

### CODE REVIEW

**Comment by lingsamuel** [comment]
The `unit_test` failed since the second commit, it complains
```diff
            - 	config modifies minikube config files using subcommands like "minikube config set driver kvm"
            + 	config modifies minikube config files using subcommands like "minikube config set driver kvm2"
```
Weird.

**Comment by afbjorklund** [comment]
> The `unit_test` failed since the second commit, it complains
> 
> ```diff
>             - 	config modifies minikube config files using subcommands like "minikube config set driver kvm"
>             + 	config modifies minikube config files using subcommands like "minikube config set driver kvm2"
> ```
> 
> Weird.

It was an "accident" with the new ISO version (f3be305abb7c609130b6957b2b63ae924113770f), I fixed it in #9941

**Comment by sharifelgamal** [comment]
If you pull master, the unit tests should pass. I'll take a closer look at this PR soon, thanks for all the hard work put into it. This is a great feature to add to multinode.

**Comment by lingsamuel** [comment]
Rebased.

**Comment by lingsamuel** [comment]
Tests failed because I use kubeadm `--upload-certs` and `init phase upload-certs --upload-certs` to make `join --control-plane` works.

Changelog: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.15.md
`--control-plane` and `--upload-certs` introduced in 1.15.

**Comment by lingsamuel** [comment]
Some macOS test timeout due to:
> Hosted macOS workflows will experience longer wait times during the week of December 14th. For more information, see actions/virtual-environments#2247

**Comment by medyagh** [comment]
thank you for this PR, I will make sure it gets added to the to be reviewed list.

**Comment by lingsamuel** [comment]
We need kubeadm 1.15 to make Jenkins tests work, how can I update it?

**Comment by medyagh** [comment]
I like this PR but for this release I think this is gonna add complexities.

**Comment by sadlil** [comment]
I am going to take a look at this PR to review. While I am doing that I have a few points to discuss:

1. Why only support in `node add` why not also support something like `minikube start --control-plane=2 --node=5`. 

2. If a minikube cluster is not started with multi node (with --node) flag there is some problem with CNI and multi node networking. See - https://github.com/kubernetes/minikube/issues/9838. Perhaps we need to have a check for this as well when doing multi control plane.

**Comment by lingsamuel** [comment]
> 1. Why only support in `node add` why not also support something like `minikube start --control-plane=2 --node=5`.
> 2. If a minikube cluster is not started with multi node (with --node) flag there is some problem with CNI and multi node networking. See - #9838. Perhaps we need to have a check for this as well when doing multi control plane.

1. because I forget it :joy: will add soon
2. I noticed this issue, and already tested with #9875. Pods get unique IPs.

File: test/integration/multinode_test.go [hunk 1]
```diff
@@ -42,269 +50,280 @@ func TestMultiNode(t *testing.T) {
...
-
-	if strings.Count(rr.Stdout.String(), "host: Running") != 2 {
-		t.Errorf("status says both hosts are not running: args %q: %v", rr.Command(), rr.Stdout.String())
-	}
-
-	if strings.Count(rr.Stdout.String(), "kubelet: Running") != 2 {
-		t.Errorf("status says both kubelets are not running: args %q: %v", rr.Command(), rr.Stdout.String())
-	}
-
+	s.startCluster()
```
Comment (tstromberg) [inline test/integration/multinode_test.go:101]: First, the implementation code looks great. My only hesitation is with this integration test: Could you please remove these helper methods and in-line the test code instead? 

Readers should need to read only the test body to understand the test. My advice is not sacrifice test clarity in the interest of removing code duplication. People only read tests when they are broken, so lets make them as simple as possible.

I would also suggest removing the *clusterStatus argument as well, if possible.

Thank you so much for your contribution! I can't wait to see it merged.

Comment (lingsamuel) [reply test/integration/multinode_test.go:101]: IMO helper functions will make things more clear. In this situation, it's hard to set these constants right, especially they affect each other across so many functions.
What's more, we need a status variable to make functions like `stop/start/delete` reusable.

**Comment by lingsamuel** [comment]
Rebased.

**Comment by lingsamuel** [comment]
Add a k8s version check, this may let tests pass.
image

File: test/integration/multinode_test.go (hunk 1 above)
Comment (lingsamuel) [reply test/integration/multinode_test.go:101]: /ping
what's your opinion @medyagh

**Comment by medyagh** [comment]
@lingsamuel thanks for updating this, I still need to verify if this PR could be accepted in minikube's vision without affection the code complexity. will do a discuss on next office hours if you like to attend on monday please we could discuss it

**Comment by lingsamuel** [comment]
I am afraid i can't participant.

**Comment by medyagh** [comment]
@lingsamuel thanks for this PR, at this release we dont have the bandwith to accept this PR but we eventually we can accept this.
the main concerns the use case for a local developement is hard to emulate, the end user would need to deploy a loadbalacner.

**Comment by medyagh** [comment]
@lingsamuel update: we discussed this in the triage meeting, we like to re-introduce this PR again after we deliver the minikube auto pause feature.  https://github.com/kubernetes/minikube/issues/10326

**Pull request closed without merging**